### PR TITLE
Rework manual Connection close in C#

### PR DESCRIPTION
**Pull request by bernardnormier**
Revision: 2

This PR changes the API an application uses to manually close a Connection in C#.
This API was and remains only for application use; it's not used internally by the Ice C# core.

See #2624 and #2652. I intend to perform similar refactoring in other language mappings if this PR is approved.

Ice 3.7 API:
```cs
public enum ConnectionClose { Forcefully, Gracefully, GracefullyWithWait }

public interface Connection
{
    void close(ConnectionClose mode);
    ...
}
```

New API:
```cs
public interface Connection
{
        /// <summary>
        /// Aborts this connection.
        /// </summary>
        void abort();

        /// <summary>
        /// Closes the connection gracefully after waiting for all outstanding invocations to complete.
        /// </summary>
        /// <returns>A task that completes when the connection is closed.</returns>
        /// <remarks>If this operation takes longer than the configured close timeout, the connection is aborted with a
        /// <see cref="CloseTimeoutException"/>.</remarks>
        Task closeAsync();
}
```

The new API is partially compatible with the old API.

| With 3.7 API                                                                   | With new  API             | Remarks |
| --------------------------------------------------- |----------------------|----------|
| `connection.close(ConnectionClose.Forcefully);`    | `connection.abort();` | exact match |
| `connection.close(ConnectionClose.Gracefully);`    | `_ = connection.closeAsync();` or `await connection.closeAsync();`  | partial match |
| `connection.close(ConnectionClose.GracefullyWithWait);`    |  `_ = connection.closeAsync();` or `await connection.closeAsync();` | partial match |

The 3.7 ability to close(Gracefully) without waiting for outstanding invocations is gone. It was an odd option since any outstanding invocation would be aborted in this case, which makes the closure not that graceful. It made sense only when there was no outstanding invocation - and in this case Gracefully and GracefullyWithWait are equivalent.

Also new in this PR: `closeAsync` is now bounded by the configured close timeout.

### CODE REVIEW

File: csharp/src/Ice/ConnectionI.cs [hunk 1]
```diff
@@ -2906,6 +2945,7 @@ internal void completed(LocalException ex)
     private ConnectionInfo _info;
 
     private CloseCallback _closeCallback;
+    private readonly TaskCompletionSource _closed = new(); // can run synchronously
```
Comment (bernardnormier) [inline csharp/src/Ice/ConnectionI.cs:2952]: See code where we call SetResult/SetException. The synchronous completion is consistent with the way we call the close callback, but maybe I am missing a potential pitfall?

File: csharp/src/Ice/Connection.cs [hunk 2]
```diff
@@ -50,39 +52,20 @@ public ConnectionInfo(ConnectionInfo underlying, bool incoming, string adapterNa
...
-    {
-        /// <summary>
-        /// Manually close the connection using the specified closure mode.
-        /// </summary>
-        /// <param name="mode">Determines how the connection will be closed.
-        ///  </param>
-        void close(ConnectionClose mode);
+        /// <param name="waitForInvocations">If true, this operation waits for all pending invocations to complete
+        /// before closing the connection.</param>
+        /// <returns>A task that completes when the connection is closed.</returns>
```
Comment (externl) [inline csharp/src/Ice/Connection.cs:67]: What does it mean for an incoming connection?

File: csharp/src/Ice/Connection.cs [hunk 3]
```diff
@@ -50,39 +52,20 @@ public ConnectionInfo(ConnectionInfo underlying, bool incoming, string adapterNa
...
+        Task closeAsync(bool waitForInvocations = true);
```
Comment (externl) [inline csharp/src/Ice/Connection.cs:68]: I guess we're sticking with camelCase for consistency with the rest of the API?

**Review by externl — APPROVED** [review]
Looks good to me! I like this new API.

File: csharp/src/Ice/Connection.cs (hunk 2 above)
Comment (bernardnormier) [reply csharp/src/Ice/Connection.cs:67]: It's the same for all connections.

You can send outgoing requests on an incoming connection. That's what we call "bidir".

File: csharp/src/Ice/Connection.cs (hunk 3 above)
Comment (bernardnormier) [reply csharp/src/Ice/Connection.cs:68]: Yes.

File: csharp/src/Ice/Connection.cs (hunk 2 above)
Comment (externl) [reply csharp/src/Ice/Connection.cs:67]: I mean, what if I get the connection of an incoming request and maybe then or some point later call closeAsync. Does close take into account ongoing dispatches for that connection?

**Review by externl — APPROVED** [review]

File: csharp/src/Ice/Connection.cs (hunk 2 above)
Comment (bernardnormier) [reply csharp/src/Ice/Connection.cs]: I thought close[Async] already waited for dispatches to complete (once state = StateClosing), but after looking at the code, I don't see how we do it.

In particular:
https://github.com/zeroc-ice/ice/blob/106c9299443791bced745eacfd5d6859505502ad/csharp/src/Ice/ConnectionI.cs#L267

waits for upcallCount (= work items in the thread pool), not dispatches. And it's not clear to me if/where we call this waitUntilFinished when closing a connection.

Maybe @bentoi can shed light on this logic?

Now, in IceRPC (which is maybe what you're thinking about), ShutdownAsync waits for both invocations and dispatches to complete _after_ refusing new invocations:

https://github.com/icerpc/icerpc-csharp/blob/b1f65d8d66b671c9948c7bfb04c3881243ba6a96/src/IceRpc/Internal/IceProtocolConnection.cs#L476
https://github.com/icerpc/icerpc-csharp/blob/b1f65d8d66b671c9948c7bfb04c3881243ba6a96/src/IceRpc/Internal/IceProtocolConnection.cs#L502

It's nicer but not trivial to retrofit in Ice.

Also, with IceRPC, the application cannot grab some connection and shut it down or abort it. You can only shut down a connection indirectly by shutting down its container (ClientConnection, ConnectionCache, Server). Same for abort (DestroyAsync in IceRPC).

Comment (bentoi) [reply csharp/src/Ice/Connection.cs]: > And it's not clear to me if/where we call this waitUntilFinished when closing a connection.

It's used to implement `ObjectAdapter::waitForDeactivate` and `Communicator::waitForShutdown`. These methods return once all the connection "activity" is done and this includes application code such as dispatch completion and callbacks (close callback, AMI callbacks, etc).

File: csharp/src/Ice/Connection.cs [hunk 4]
```diff
@@ -2,6 +2,8 @@
 
 #nullable enable
 
+using Ice.Internal;
```
Comment (InsertCreativityHere) [inline csharp/src/Ice/Connection.cs:5]: Why do we need this `using` now?
We didn't have it before, and it doesn't seem like any internal types were added here.

**Review by InsertCreativityHere — APPROVED** [review]

File: csharp/src/Ice/ConnectionI.cs [hunk 5]
```diff
@@ -155,44 +155,62 @@ internal void destroy(int reason)
...
+                            closedByApplication: true));
+                    break; // exit the forever loop
                 }
+                else
+                {
+                    Debug.Assert(waitForInvocations && _asyncRequests.Count > 0);
+                    if (_asyncRequestsCompleted is null || _asyncRequestsCompleted.Task.IsCompleted)
+                    {
+                        // Create or recreate the task completion source within lock
+                        _asyncRequestsCompleted = new TaskCompletionSource();
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/ConnectionI.cs]: I removed this code - this comment is no longer applicable.

File: csharp/src/Ice/Connection.cs (hunk 2 above)
Comment (bernardnormier) [reply csharp/src/Ice/Connection.cs]: In other words, it's not called when a specific connection is closed manually using `Connection::closeAsync`?

File: csharp/src/Ice/Connection.cs (hunk 4 above)
Comment (bernardnormier) [reply csharp/src/Ice/Connection.cs:5]: Removed. It was not needed.

File: csharp/src/Ice/ConnectionI.cs [hunk 6]
```diff
@@ -155,44 +155,38 @@ internal void destroy(int reason)
...
+            if (_state < StateClosing)
+            {
+                if (_asyncRequests.Count == 0)
                 {
-                    Monitor.Wait(this);
+                    doApplicationClose();
+                }
+                else
+                {
+                    _closeRequested = true;
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/ConnectionI.cs:182]: I tried but I don't see how to do it.

The issue is ConnectionI::finish executes without locking this and completes _closed. And we want finish to complete _closed even if the application calls Connection::closeAsync on a closed connection.

**Comment by bernardnormier** [comment]
> Removing the close method sounds good as its semantics are really unclear. I believe it was primarily intended to be used on the client side to interrupt pending client invocations and gracefully close the connection...
> 
> As you mentioned, it would be better to have semantics closer to the IceRPC shutdown method:
> 
> * prevent new invocations
> * wait for upcall count == 0 (i.e. wait for dispatch, invocations, other callbacks to complete)

One difficulty here is upcall does _not_ correspond to dispatch/invocation. It corresponds to work items to be executed by the thread pool. So upcall count can be 0 even though we have outstanding invocations and dispatches.

Maybe we could look how to delay the connection closure (finish) until after all dispatches have completed in a follow-up PR?

> 
> One difference would be that it wouldn't refuse new incoming requests, I doubt we can find a nice solution for this.

What about refusing new dispatches? Can we do that with Ice? (We do it with IceRPC).

> 
> It would be good to also consider the semantics of communicator/object adapter `DestroyAsync`, `ShutdownAsync` and `DeactivateAsync` methods if we're going to add these async APIs. They are closely related...

They are not directly related since none of these methods call Connection::closeAsync. Connection::closeAsync is purely for application consumption.

But more generally, yes, updating Connection::close is just a first step for #2624.

> 
> So if we remove `Connection::close`, we should consider removing the object adapter/communicator deactivate, waitForDeactivite, shutdown, waitForShutdown methods...

**Comment by bernardnormier** [comment]
> > What about refusing new dispatches? Can we do that with Ice? (We do it with IceRPC).
> How would you report it to the client? That's the main difficulty.

The logic should be the same IceRPC with the ice protocol, i.e. we discard this incoming request to avoid delaying the graceful closure.

We don't report this discarding with IceRPC(ice), at least not directly.  For a two-way invocation, if the client:
 - gets a reply, it means the dispatch completed before the connection shutdown
 - gets a IceRpcException(InvocationCanceled) (CloseConnectionException with Ice), it means the dispatch was discarded by the shutdown
 
For a oneway invocation, the client can't distinguish between a successful oneway and a discarded oneway. But that's the general nature of oneways.

**Comment by bernardnormier** [comment]
> > So upcall count can be 0 even though we have outstanding invocations and dispatches.

> Why is this is a big issue that would require making the code more complicated?

upcalls is an implementation detail. Users should not think about upcalls. invocations and dispatches are application-level concepts.

And related to the previous comment, it would be nicer if shutdown waited for all invocations and dispatches to complete. It actually already waits for two-way invocations to complete in a round-about manner, since closeAsync:
 - first waits for all two-way invocations to complete
 - then initiate closure
 - as soon as closure is initiated, new invocations are rejected: https://github.com/zeroc-ice/ice/blob/64a486acd32a04f1af431c4d36af9dd0acd08544/csharp/src/Ice/ConnectionI.cs#L319

**Pull request merged**